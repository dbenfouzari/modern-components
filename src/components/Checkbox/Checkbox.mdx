---
name: Checkbox
---

import { Playground, PropsTable } from 'docz';
import Checkbox from './';

# Checkbox

<PropsTable of={Checkbox} />

## Basic usage

### Unchecked

<Playground>
    <Checkbox value={false} onChange={nextValue => console.log(nextValue)}>
        Hello World!
    </Checkbox>
</Playground>

### Checked

<Playground>
    <Checkbox value={true} onChange={nextValue => console.log(nextValue)}>
        Hello World!
    </Checkbox>
</Playground>

### Disabled

<Playground>
    <Checkbox disabled={true} value={true} onChange={nextValue => console.log(nextValue)}>
        Hello World!
    </Checkbox>
</Playground>
