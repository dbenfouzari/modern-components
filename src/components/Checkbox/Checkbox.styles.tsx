import { mix } from "polished";
import styled from "../../styled-components";

export const CheckboxWrapper = styled.span``;
export const StyledCheckbox = styled.input`
  position: absolute;
  left: -9999px;

  &:disabled + label:before {
    cursor: not-allowed;
    background-color: ${props => mix(0.6, "#eee", props.theme.greenSea)};
  }

  &:disabled + label:after {
    cursor: not-allowed;
  }

  & + label {
    position: relative;
    padding-left: 25px;
    cursor: pointer;

    &:before {
      content: "";
      position: absolute;
      left: 0;
      top: 2px;
      width: 17px;
      height: 17px;
      border: 1px solid ${props => props.theme.peterRiver};
      border-radius: 3px;
    }

    &:after {
      content: "âœ”";
      position: absolute;
      top: 0;
      left: 4px;
      font-size: 19px;
      color: ${props => props.theme.peterRiver};
      transition: all 0.2s ease;
    }
  }

  &:not(:checked) + label:after {
    opacity: 0;
    transform: scale(0);
  }

  &:checked + label:after {
    opacity: 1;
    transform: scale(1);
  }
`;
