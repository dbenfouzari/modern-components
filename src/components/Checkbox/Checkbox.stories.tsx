import { StateDecorator, Store } from "@sambego/storybook-state";
import { action } from "@storybook/addon-actions";
import { storiesOf } from "@storybook/react";
import * as React from "react";

import Checkbox from ".";

storiesOf("Checkbox", module)
  .add("unchecked", () => {
    const handleChange = (nextValue: boolean) => {
      action("checkboxChange")(nextValue);
    };

    return (
      <Checkbox value={false} onChange={handleChange}>
        Hello World!
      </Checkbox>
    );
  })
  .add("checked", () => {
    const handleChange = (nextValue: boolean) => {
      action("checkboxChange")(nextValue);
    };

    return (
      <Checkbox value={true} onChange={handleChange}>
        Hello World!
      </Checkbox>
    );
  })
  .add("disabled", () => {
    const handleChange = (nextValue: boolean) => {
      action("checkboxChange")(nextValue);
    };

    return (
      <Checkbox disabled={true} value={true} onChange={handleChange}>
        Hello World!
      </Checkbox>
    );
  });
