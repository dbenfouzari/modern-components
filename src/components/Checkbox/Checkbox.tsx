import * as React from "react";
import { CheckboxWrapper, StyledCheckbox } from "./Checkbox.styles";

interface InputProps {
  /**
   * @since 0.0.1
   * @version 0.0.1
   *
   * Value passed to the checkbox
   */
  value: boolean;
  /**
   * @since 0.0.1
   * @version 0.0.1
   *
   * The action that will be called when user checks/unchecks
   */
  onChange: (nextValue: boolean) => void;
  /**
   * @since 0.0.1
   * @version 0.0.1
   *
   * What will be aside to checkbox
   */
  children?: string;
  /**
   * @since 0.0.1
   * @version 0.0.1
   *
   * Is it disabled ?
   */
  disabled?: boolean;
}

/**
 * @since 0.0.1
 * @version 0.0.1
 * @author Donovan BENFOUZARI <d.benfouzari@gmail.com>
 *
 * This is a simple input
 */
const Input = ({ value, onChange, children, disabled }: InputProps) => {
  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    console.log(":)");
    onChange(event.currentTarget.checked);
  };

  return (
    <CheckboxWrapper>
      <StyledCheckbox
        disabled={disabled}
        type="checkbox"
        onChange={handleChange}
        defaultChecked={value}
      />
      <label>{children}</label>
    </CheckboxWrapper>
  );
};

export default Input;
