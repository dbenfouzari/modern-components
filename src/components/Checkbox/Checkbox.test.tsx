import { shallow } from "enzyme";
import * as React from "react";
import renderer from "react-test-renderer";
import { ThemeProvider } from "../../styled-components";
import defaultTheme from "../../theme";
import Checkbox from "./Checkbox";

const noop = () => null;

describe("Checkbox", () => {
  it("renders correctly", () => {
    const wrapper = renderer
      .create(
        <ThemeProvider theme={defaultTheme}>
          <Checkbox value={true} onChange={noop}>
            Coucou
          </Checkbox>
        </ThemeProvider>
      )
      .toJSON();

    expect(wrapper).toMatchSnapshot();
  });
});
