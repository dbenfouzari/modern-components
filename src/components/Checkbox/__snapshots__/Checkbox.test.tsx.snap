// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders correctly 1`] = `
.c0 {
  position: absolute;
  left: -9999px;
}

.c0 + label {
  position: relative;
  padding-left: 25px;
  cursor: pointer;
}

.c0 + label:before {
  content: "";
  position: absolute;
  left: 0;
  top: 2px;
  width: 17px;
  height: 17px;
  border: 1px solid #28bcd6;
  border-radius: 3px;
}

.c0 + label:after {
  content: "âœ”";
  position: absolute;
  top: 0;
  left: 4px;
  font-size: 19px;
  color: #28bcd6;
  -webkit-transition: all 0.2s ease;
  transition: all 0.2s ease;
}

.c0:not(:checked) + label:after {
  opacity: 0;
  -webkit-transform: scale(0);
  -ms-transform: scale(0);
  transform: scale(0);
}

.c0:checked + label:after {
  opacity: 1;
  -webkit-transform: scale(1);
  -ms-transform: scale(1);
  transform: scale(1);
}

<span
  className=""
>
  <input
    checked={true}
    className="c0"
    onChange={[Function]}
    type="checkbox"
  />
  <label>
    Coucou
  </label>
</span>
`;
